Feature: Implement API

CeNTePeDe opened the pull request.

CeNTePeDe committed:
created
- serializer
- router
- ProductDAO
- data_base module
added
- User-Agent to constant variables

slavadan commented on Dockerfile:
```diff
@@ -0,0 +1,13 @@
+RUN pip install pipenv && pipenv install --system
+
+COPY . .
+
```
дубликат убери, я же тебе сам даже поправил

slavadan commented on data_base/database.py:
```diff
@@ -0,0 +1,35 @@
+            logger.info(f"products {item}")
+            product_item = serializer_products_by_url(item)
+            return product_item
+
```
неправильная реализация DAO класса. почитай что такое DAO class. Заодно повтори SOLID, MVC(есть по заданию)

slavadan commented on main.py:
```diff
@@ -1,5 +1,27 @@
+from config import SettingsDev, SettingsProd, SettingsStage
 
-app = FastAPI()
+from routers.product import routers
```
плохие импорты, я по моему говорил про настройку black,  flake добавь еще mypy

slavadan commented on parsing_function/parse_product.py:
```diff
@@ -0,0 +1,121 @@
+def gather_data_to_list(product_data: list[dict]):
+    global DATA
+    DATA += product_data
+    return DATA
```
обсуждали почему не испоьзовать глобальную переменную для таких целей. Даже если ты решила использовать эту переменную, как думаешь, что будет с ней если вызвать функцию парсера 20 раз

slavadan commented on routers/product.py:
```diff
@@ -0,0 +1,21 @@
+    logger.info("get url")
+    product_dao.insert_many_products(url)
+    logger.info("collection is inserted")
+    return product_dao.get_all_products(url)
```
ветка с названием feature API но из апи тут только один метод. Тоже уже сколько раз обговаривалось. 1 фича/исправление = 1 ветка. У тебя незвание ветки и pr не соответствует тому, что ты делаешь

slavadan commented on schema/schemas.py:
```diff
@@ -0,0 +1,22 @@
+        "url": data["url"],
+        "products": list_product_serializer(data["products"])
+    }
+
```
почему у тебя функции имеют имя существительное?

CeNTePeDe committed:
added:
- exception_handler
- combined_list_of_products to gather_data

slavadan commented on controllers/exception_handler.py:
```diff
@@ -0,0 +1,11 @@
+        status_code = 404
+        detail = "Invalid URL or not found"
+    return HTTPException(status_code=status_code, detail=detail)
+
```
хэндлер != контроллер(читай внимательней что такое controller), хотя он может считаться частью контроллера

slavadan replied:
+ хэндлер ловит у тебя все ошибки, но делает проверку AttribeError = бред

slavadan commented on routers/product.py:
```diff
@@ -0,0 +1,24 @@
+        product_dao.insert_products(url, products)
+        return product_dao.get_products(url)
+    except Exception as e:
+        raise handle_exception(e)
```
неправильно использование handle_exception

slavadan commented:
Реши мерж конфликты

CeNTePeDe committed:
deleted exception_handler.py
renamed:
- exception_404_handler to InvalidUrlInputError
added error_handler to gather_data function

slavadan commented on core/exception.py:
```diff
@@ -0,0 +1,14 @@
+
+def exception_404_handler(request: Exception, exc: InvalidUrlInputError):
+    return JSONResponse(status_code=status.HTTP_404_NOT_FOUND,
+                        content={"message": f"{exc.name} cannot be found."})
```
хэндлеры в отдельном файле

CeNTePeDe committed:
conflicts were resoved

CeNTePeDe committed:
created file exception_handler.py

slavadan commented:
А что-то кроме одного API метода для запуска парсинга еще будет?

CeNTePeDe committed:
added:
-get_filter_products, delete_products to ProductDAO
-get_filter_products, delete_collection to controller

slavadan commented on main.py:
```diff
@@ -12,3 +24,6 @@
 
 app = FastAPI(title=settings.PROJECT_NAME)
 
+app.add_exception_handler(Exception, exception_404_handler)
```
тут забыла на свой эксепшн поменять

slavadan commented on parsing_function/parse_product.py:
```diff
@@ -5,20 +5,21 @@
+from core.constant_variables import (MAIN_PAGE,
+                                     HEADERS,
+                                     NUMBER_OF_PRODUCT_PER_PAGE)
+from core.exception import InvalidUrlInputError
```
настрой ты уже себе этот несчастный black и isort

slavadan commented on parsing_function/parse_product.py:
```diff
@@ -86,25 +88,31 @@ async def get_data_for_each_product(
-        pages_count = math.ceil(number_of_products / 60)
+        number_of_products_locator = soup.find("span", class_="d-catalog-header__product-counter")
+        if number_of_products_locator is None:
+            raise InvalidUrlInputError(name=url)
```
отступ после if. старайся разделять код по логике чтоле с отступами. А то у тебя пластом все лежит

slavadan commented on routers/product.py:
```diff
@@ -0,0 +1,32 @@
+@routers.delete("/delete", status_code=status.HTTP_204_NO_CONTENT)
+def delete_collection(url: AnyUrl):
+    product_dao.delete_products(url)
+    return {"message": "Document was deleted"}
```
я вроде просил crud сделать)

slavadan commented on main.py:
```diff
@@ -1,9 +1,21 @@
-from settings import init_settings
+from config import SettingsDev, SettingsProd, SettingsStage
+from core.exception_handler import exception_404_handler
+from routers.product import routers
```
та настрой уже black и isort нестчастные

slavadan commented on data_base/database.py:
```diff
@@ -0,0 +1,48 @@
+                if min_price < float(product_item["price"].split()[0]) < max_price:
+                    sorted_product_list.append(product_item)
+                    logger.info(f"{sorted_product_list}")
+            return serialize_list_of_products(sorted_product_list)
```
будто бы первый раз фильтры свои пишешь(или ты в джанго не писала?). Фильтры отдельным классом. Загуглила хотябы). Забыла -> указать что возвращает функция

slavadan commented on data_base/database.py:
```diff
@@ -0,0 +1,48 @@
+                    logger.info(f"{sorted_product_list}")
+            return serialize_list_of_products(sorted_product_list)
+
+    def delete_products(self, url: AnyUrl):
```
забыла указать, что возвращает функция

slavadan commented on routers/product.py:
```diff
@@ -0,0 +1,32 @@
+    logger.info("get url")
+    products = asyncio.run(gather_data(url))
+    product_dao.create_products(url, products)
+    return product_dao.get_products(url)
```
только что заметил, что у тебя метод называется get_***, но по факту это post. несостыковочка

CeNTePeDe committed:
created
- ProductFilter
- post_products
- get_products

slavadan commented on routers/product.py:
```diff
@@ -0,0 +1,37 @@
+
+
+@routers.post("/", status_code=status.HTTP_201_CREATED)
+def post_products(url: AnyUrl) -> list[dict]:
```
start_parsing_task

slavadan commented on routers/product.py:
```diff
@@ -0,0 +1,37 @@
+@routers.delete("/delete", status_code=status.HTTP_204_NO_CONTENT)
+async def delete_collection(url: AnyUrl):
+    product_dao.delete_products(url)
+    return {"message": "Document was deleted"}
```
не хватает методов для круда

slavadan commented on data_base/database.py:
```diff
@@ -0,0 +1,41 @@
+from pymongo.collection import Collection
+
+from schema.schemas import (serialize_list_of_products,
+                            serialize_products_with_url)
```
(
serialize_list_of_products,
serialize_products_with_url
)

slavadan commented on core/exception_handler.py:
```diff
@@ -0,0 +1,11 @@
+from core.exception import InvalidUrlInputError
+
+
+def exception_404_handler(request: Exception, exc: InvalidUrlInputError):
```
-> JSONResponse

slavadan commented on core/constant_variables.py:
```diff
@@ -1,6 +1,12 @@
-MAIN_PAGE_MEN = "https://www.lamoda.by/men-home/"
-MAIN_PAGE_WOMAN = "https://www.lamoda.by/women-home/"
-MAIN_PAGE_KIDS = "https://www.lamoda.by/kids-home/"
+NUMBER_OF_PRODUCT_PER_PAGE = 60
```
: int

CeNTePeDe committed:
created:
- BaseDAO class
- CategoryDAO
- ProductDAO
- category_routers
- product_routers
added to schemas:
- serialize_category
- serialize_list_of_category

slavadan commented on .gitignore:
```diff
@@ -15,3 +15,4 @@ __pycache__
 /.idea/.gitignore
 /.env
 
+/.pre-commit-config.yaml
```
убрать

slavadan commented on database/connection_categories.py:
```diff
@@ -0,0 +1,48 @@
+    def get(self):
+        categories = list(self.collection.find())
+        logger.info(f"all categories {categories}")
+        return categories
```
list[Category]

slavadan commented on database/connection_categories.py:
```diff
@@ -0,0 +1,48 @@
+        if result.modified_count > 0:
+            return {"message": "Update document"}
+        else:
+            return {"message": "Document not found"}
```
лишний else

slavadan commented on database/connection_categories.py:
```diff
@@ -0,0 +1,48 @@
+            logger.info(f"category look like this {category_dict}")
+            category = self.collection.insert_one(category_dict)
+            logger.info(f"category inserted {category.inserted_id}")
+            return category.inserted_id
```
вместо дикта сюда должен приходить Category

slavadan commented on database/connection_categories.py:
```diff
@@ -0,0 +1,48 @@
+        else:
+            return {"message": "Document not found"}
+
+    def delete(self, category_name):
```
тайп хинты

slavadan commented on database/connection_products.py:
```diff
@@ -0,0 +1,43 @@
+from core.base_class import AbstractDAO
+
+logger = logging.getLogger(__name__)
+CLIENT = MongoClient("mongodb://user:1111@mongo:27017")
```
lower case

slavadan commented on database/connection_products.py:
```diff
@@ -0,0 +1,43 @@
+    def __init__(self):
+        self.collection = DB["products"]
+
+    def get(self):
```
тайпхинты

slavadan commented on database/connection_products.py:
```diff
@@ -0,0 +1,43 @@
+        if deleted_product.deleted_count == 1:
+            return {"message": "Product is deleted"}
+        else:
+            return {"message": "Document not found"}
```
лишний else

slavadan commented on models/product_models.py:
```diff
@@ -0,0 +1,15 @@
+class ProductModel(BaseModel):
+    name_product: str
+    picture_link: AnyUrl
+    price: float
```
не используй float

slavadan commented on routers/product_routers.py:
```diff
@@ -0,0 +1,55 @@
+async def create_product(product: ProductModel):
+    product_dict = dict(product)
+    category_dict = dict(product_dict.pop("category"))
+    id_category = category_dao.create_category(category_dict)
```
передавать модель

CeNTePeDe committed:
added Decimal field
product_routers:
- get_product
- create_product
- update_product
- delete_product
added
- pre-commit-config.yaml
- methods: get_all_item, get_item to AbstractDAO
category_routers added:
- get_categories
- get_category
- update_category
- create_category
- delete_category

slavadan commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,18 @@
+-   repo: https://github.com/pycqa/isort
+    rev: 5.12.0
+    hooks:
+    -   id: isort
```
пустая линия в конце файла

slavadan commented on database/connection_categories.py:
```diff
@@ -0,0 +1,52 @@
+
+    def get_all_item(self) -> Any:
+        categories = self.collection.find()
+        return categories
```
:(

slavadan commented on database/connection_categories.py:
```diff
@@ -0,0 +1,52 @@
+            logger.info(f"category look like this {category_dict}")
+            category = self.collection.insert_one(category_dict)
+            logger.info(f"category inserted {category.inserted_id}")
+            return category.inserted_id
```
лишний else, что-то category и category_dict названия меня напрягают.

slavadan commented on database/connection_categories.py:
```diff
@@ -0,0 +1,52 @@
+    def get_item(self, category_id: str) -> Mapping[str, Any]:
+        cat_id = ObjectId(category_id)
+        category = self.collection.find_one({"_id": {"$eq": cat_id}})
+        return category
```
:(

slavadan commented on database/connection_products.py:
```diff
@@ -0,0 +1,48 @@
+        super().__init__(self.collection)
+
+    def get_all_item(self) -> Cursor:
+        return self.collection.find()
```
:((((

slavadan commented on database/connection_products.py:
```diff
@@ -0,0 +1,48 @@
+    def get_item(self, product_id: str) -> Mapping[str, Any]:
+        prod_id = ObjectId(product_id)
+        product = self.collection.find_one({"_id": {"$eq": prod_id}})
+        return product
```
:(

slavadan commented on database/connection_products.py:
```diff
@@ -0,0 +1,48 @@
+        price_old = str(product.pop("price"))
+        product["price"] = str(price_old)
+        product = self.collection.insert_one(product)
+        return product.inserted_id
```
product имеет тип Any

slavadan commented on core/base_class.py:
```diff
@@ -0,0 +1,30 @@
+        pass
+
+    @abstractmethod
+    def delete_item(self, _id):
```
вместо pass, raise NotImplementedError

CeNTePeDe committed:
removed serializers and return Pydentic model
base_class added:
- NotImplementationError
category/product routers added:
- Exceptions
